### ISSUE TITLE
Detect if image is HEIC or AVIF

### ISSUE THREAD

**Issue by dlemstra** [strukturag/libheif#427]
The @ImageMagick library is using this library and we would like to know what would be the best way to detect if the image is either `HEIC` or `AVIF`. We have two different files with the following header:

```
// FILE1
00000000: 0000 0018 6674 7970 6176 6966 0000 0000 ....ftypavif....
00000010: 6176 6966 6d69 6631 0000 01fe 6d65 7461 avifmif1....meta
```

```
// FILE2
00000000: 0000 001c 6674 7970 6d69 6631 0000 0000 ....ftypmif1....
00000010: 6d69 6631 6176 6966 6d69 6166 0000 00f2 mif1avifmiaf....
```

`FILE1` has `avif` as the `heif_main_brand` but the second image has `mif1` instead. But both files contain `avif` in the compatible brands after the `heif_main_brand`. Would it be okay to also classify `FILE2` as `AVIF` because it contains `avif` in the compatible brands?

If that could be used to detect the "format" would it be possible to add something to the C API that we can use to check if the file contains a certain compatible brand?

```
int heif_has_compatible_brand(const uint8_t* data, int len, enum heif_brand brand);
```

**Comment by farindk** [comment]
I have added such an API in commit f4756c2a77d1a5f4c07c7effbbd9a1494a54dbd1.
Now, you can check for compatible brands either with (note that I have deprecated `enum heif_brand`).
````
int heif_has_compatible_brand(const uint8_t* data, int len, const char* brand_fourcc);
````
(For HEIF, you would probably check against "heic" and "heix", for AVIF only for "avif").

Or you can use
````
int heif_list_compatible_brands(const uint8_t* data, int len, heif_brand2* out_brands, int out_size);
````
and then check the returned list for brands using
````
  brandlist[i] == heif_fourcc_to_brand("avif")
````

**Comment by dlemstra** [comment]
Thanks for the quick response and change.

**Comment by dlemstra** [comment]
I was wondering if it would be allowed to do this instead:

```c
enum heif_brand: uint32_t
{
  heif_unknown_brand = 0,
  heif_heic = fourcc_to_uint32("heic"),
  heif_avif = fourcc_to_uint32("avif")
};
```

Not sure if that would break the API or ABI? Also not sure if every platform support `uint32_t` as the datatype for an enumeration. But it would probably be easier to use this API with an enumeration.

**Comment by dlemstra** [comment]
I did not realize there would be such a variaty in the brands. Taking another look at the code made me realize that we would probably only use `heif_has_compatible_brand` and check if it contains "avif". 

You might want to add a check on the length of `brand_fourcc`. If a string with a length of 1 is used in that method it will result in an out of bounds read.

And maybe `heif_list_compatible_brands` should be changed to do the allocation itself and `out_size` changed into a pointer so the user knows the size of the array instead of providing an array with an arbitrary size?

**Comment by farindk** [comment]
Thanks for the code review. I have added the improvements. The `fourcc` input checks are implemented such that it does not have to be a null-terminated 4-character string. It also may be a simple array of 4 chars. This may be handy if we just want to point into the middle of a stream without copying it to a dummy string.

**Comment by farindk** [comment]
New API is included in release v1.11.0.

**Comment by dlemstra** [comment]
Thanks!
